File: python/connect4.py
"""Pure-Python Connect4 game state for gameplay server."""

from __future__ import annotations


class Connect4:
    ROWS = 6
    COLS = 7

    def __init__(self, initial_board: list[list[int]] | None = None):
        self.board: list[list[int]] = []
        self.current_player = 1
        self._finished = False
        self._reward = 0.0

        if initial_board is None:
            self.reset()
            return

        self._validate_board_shape(initial_board)
        self.board = [
            [self._normalize_cell_value(cell) for cell in row] for row in initial_board
        ]

        player1_count = 0
        player2_count = 0
        for row in self.board:
            for cell in row:
                if cell == 1:
                    player1_count += 1
                elif cell == -1:
                    player2_count += 1

        self.current_player = 1 if player1_count == player2_count else -1
        self._finished = False
        self._reward = 0.0

        for row in range(self.ROWS):
            for col in range(self.COLS):
                cell = self.board[row][col]
                if cell != 0 and self._check_win_for_player(row, col, cell):
                    self._finished = True
                    self._reward = 1.0 if cell == self.current_player else -1.0
                    return

        board_full = all(self.board[0][col] != 0 for col in range(self.COLS))
        if board_full:
            self._finished = True
            self._reward = 0.0

    @property
    def is_terminal(self) -> bool:
        return self._finished

    def reset(self) -> None:
        self.board = [[0 for _ in range(self.COLS)] for _ in range(self.ROWS)]
        self.current_player = 1
        self._finished = False
        self._reward = 0.0

    def get_legal_actions(self) -> list[int]:
        return [col for col in range(self.COLS) if self.board[0][col] == 0]

    def step(self, action: int) -> None:
        if (
            self._finished
            or action < 0
            or action >= self.COLS
            or self.board[0][action] != 0
        ):
            raise ValueError(f"Invalid action {self._finished} {action}")

        placed_row = -1
        for row in range(self.ROWS - 1, -1, -1):
            if self.board[row][action] == 0:
                self.board[row][action] = self.current_player
                placed_row = row
                break

        if placed_row == -1:
            raise ValueError(f"Invalid action {self._finished} {action}")

        if self._check_win_for_player(placed_row, action, self.current_player):
            self._finished = True
            self._reward = 1.0
        elif not self.get_legal_actions():
            self._finished = True
            self._reward = 0.0
        else:
            self.current_player = -self.current_player

    def get_board_state(self) -> list[list[int]]:
        return [row[:] for row in self.board]

    def reward(self) -> float:
        return self._reward if self._finished else 0.0

    def _validate_board_shape(self, board: list[list[int]]) -> None:
        if len(board) != self.ROWS:
            raise ValueError(f"Board must have {self.ROWS} rows")

        for row in board:
            if len(row) != self.COLS:
                raise ValueError(f"Board must have {self.COLS} columns")

    def _normalize_cell_value(self, cell: int) -> int:
        if cell == 2:
            return -1
        if cell in (-1, 0, 1):
            return cell
        raise ValueError(f"Invalid cell value: {cell}")

    def _check_win_for_player(self, row: int, col: int, player: int) -> bool:
        return (
            self._check_direction_for_player(row, col, 1, 0, player)
            or self._check_direction_for_player(row, col, 0, 1, player)
            or self._check_direction_for_player(row, col, 1, 1, player)
            or self._check_direction_for_player(row, col, 1, -1, player)
        )

    def _check_direction_for_player(
        self, row: int, col: int, d_row: int, d_col: int, player: int
    ) -> bool:
        count = 0
        for i in range(-3, 4):
            r = row + i * d_row
            c = col + i * d_col
            in_bounds = 0 <= r < self.ROWS and 0 <= c < self.COLS
            if in_bounds and self.board[r][c] == player:
                count += 1
                if count == 4:
                    return True
            else:
                count = 0
        return False

File: gameplay_server/src/game/connect4.ts
import type Game from './game.js'

type Connect4StateJson = {
  board: number[][]
}

class Connect4 implements Game {
  private static readonly ROWS = 6
  private static readonly COLS = 7

  private board: number[][]
  private currentPlayer: number
  private finished: boolean
  private rewardValue: number

  constructor(initialBoard?: number[][]) {
    this.board = []
    this.currentPlayer = 1
    this.finished = false
    this.rewardValue = 0.0

    if (initialBoard === undefined) {
      this.reset()
      return
    }

    this.validateBoardShape(initialBoard)
    this.board = initialBoard.map((row) => row.map((cell) => this.normalizeCellValue(cell)))

    let player1Count = 0
    let player2Count = 0
    for (const row of this.board) {
      for (const cell of row) {
        if (cell === 1) player1Count += 1
        else if (cell === -1) player2Count += 1
      }
    }

    this.currentPlayer = player1Count === player2Count ? 1 : -1
    this.finished = false
    this.rewardValue = 0.0

    for (let row = 0; row < Connect4.ROWS; row += 1) {
      for (let col = 0; col < Connect4.COLS; col += 1) {
        const cell = this.board[row]?.[col]
        if (cell !== undefined && cell !== 0 && this.checkWin(row, col)) {
          this.finished = true
          this.rewardValue = cell === this.currentPlayer ? 1.0 : -1.0
          return
        }
      }
    }

    let boardFull = true
    for (let col = 0; col < Connect4.COLS; col += 1) {
      const topCell = this.board[0]?.[col]
      if (topCell === 0) {
        boardFull = false
        break
      }
    }

    if (boardFull) {
      this.finished = true
      this.rewardValue = 0.0
    }
  }

  reset(): void {
    this.board = Array.from(
      { length: Connect4.ROWS },
      () => Array<number>(Connect4.COLS).fill(0),
    )
    this.currentPlayer = 1
    this.finished = false
    this.rewardValue = 0.0
  }

  getActionSize(): number {
    return Connect4.COLS
  }

  get_legal_actions(): number[] {
    const legalActions: number[] = []
    for (let col = 0; col < Connect4.COLS; col += 1) {
      if (this.board[0]?.[col] === 0) {
        legalActions.push(col)
      }
    }
    return legalActions
  }

  step(action: number): void {
    if (
      this.finished ||
      action < 0 ||
      action >= Connect4.COLS ||
      this.board[0]?.[action] !== 0
    ) {
      throw new Error(`Invalid action ${this.finished} ${action}`)
    }

    let placedRow = -1
    const placedCol = action
    for (let row = Connect4.ROWS - 1; row >= 0; row -= 1) {
      if (this.board[row]?.[action] === 0) {
        const rowRef = this.board[row]
        if (rowRef === undefined) {
          throw new Error('Invalid board state')
        }
        rowRef[action] = this.currentPlayer
        placedRow = row
        break
      }
    }

    if (placedRow === -1) {
      throw new Error(`Invalid action ${this.finished} ${action}`)
    }

    if (this.checkWin(placedRow, placedCol)) {
      this.finished = true
      this.rewardValue = 1.0
    } else if (this.get_legal_actions().length === 0) {
      this.finished = true
      this.rewardValue = 0.0
    } else {
      this.currentPlayer = -this.currentPlayer
    }
  }

  get_board_state(): Connect4StateJson {
    const serializedBoard = this.board.map((row) =>
      row.map((cell) => {
        if (cell === -1) return 2
        return cell
      }),
    )

    return { board: serializedBoard }
  }

  get_current_player(): number {
    return this.currentPlayer
  }

  is_terminal(): boolean {
    return this.finished
  }

  reward(): number {
    if (!this.finished) return 0.0
    return this.rewardValue
  }

  private validateBoardShape(board: number[][]): void {
    if (board.length !== Connect4.ROWS) {
      throw new Error(`Board must have ${Connect4.ROWS} rows`)
    }

    for (const row of board) {
      if (row.length !== Connect4.COLS) {
        throw new Error(`Board must have ${Connect4.COLS} columns`)
      }
    }
  }

  private normalizeCellValue(cell: number): number {
    if (cell === 2) return -1
    if (cell === -1 || cell === 0 || cell === 1) return cell
    throw new Error(`Invalid cell value: ${cell}`)
  }

  private checkWin(row: number, col: number): boolean {
    return (
      this.checkDirection(row, col, 1, 0) ||
      this.checkDirection(row, col, 0, 1) ||
      this.checkDirection(row, col, 1, 1) ||
      this.checkDirection(row, col, 1, -1)
    )
  }

  private checkDirection(row: number, col: number, dRow: number, dCol: number): boolean {
    let count = 0
    for (let i = -3; i <= 3; i += 1) {
      const r = row + i * dRow
      const c = col + i * dCol
      const inBounds = r >= 0 && r < Connect4.ROWS && c >= 0 && c < Connect4.COLS
      if (inBounds && this.board[r]?.[c] === this.currentPlayer) {
        count += 1
        if (count === 4) return true
      } else {
        count = 0
      }
    }
    return false
  }
}

export default Connect4
